DIPs/contractsDIP-ZT.md

```diff
@@ -0,0 +1,164 @@
+
+## Description
+### Preliminary
+
+The main proposal requires the introduction of the following three syntactic innovations:
+
+#### 1. allow `in` and `out` statements with semicolons
+
+Expand the grammar of `in` and `out` statements to include regular statements in addition to bracketed block statements:
+```
+in assert(x);
+out(r) assert(r);
+```
+This innovation preserves the basic idea of contracts as statements. Added to D, it would arguably be an improvement by itself, as many programmers are deterred from writing one-line contracts simply because of the brackets and formatting they require. 
+
+#### 2. allow multiple `in` statements
+
+The compiler groups all `in` statements, which must be consecutive, into a single contract. By itself, this seems unnecessary. But with unbracketed statements, readability can actually improve, allowing: 
+```
+in assert(x);
+in assert(y);
+out(r) assert(r);
+```
+instead of:
+```
+in {
+    assert(x);
+    assert(y);
+}
+out(r) assert(r);
+```
+#### 3. allow contracts within the function body
+
+They must occur at the beginning of the function, and act as syntax sugar for external contracts. Virtual interface functions cannot use this syntax, as the existence of contracts in a function body implies the existence of that body.  To generate `.di` files and documentation, rewrite the contracts using the old style.
+
+Note that it's possible to allow innovations 1 and 2 only within the context of innovation 3.
+
+### Proposal
+
+Combining innovations 1, 2, and 3 above, this example of current D:
+```
+int noZeroAdd(int a, int b)
+in {
+    assert(a);
+    assert(b);
+}
+out(r) {
+    assert(r);
+}
+body {
+    return a + b;
+}
+```
+turns into this:
+```
+int noZeroAdd(int a, int b) {
+    in assert(a);
+    in assert(b);
+    out(r) assert(r);
+
+    return a + b;
+}
+```
+The more familiar use with brackets is still allowed, of course, even in the body:
+```
+int noZeroAdd(int a, int b) {
+    in {
+        assert(a);
+        assert(b);
+    }
+    out(r) {
+        assert(r);
+    }
+
+    return a + b;
+}
+```
+## Alternatives
+
+Any of the three innovations above can be combined separately to produce alternatives. In particular, if innovation 3 is deemed too extravagant, innovation 1, or the combination of innovations 1 and 2, may still be desirable.
+
+Two more innovations have briefly been sketched in the forums:
+
+#### 4. [allow contracts as naked expressions](http://forum.dlang.org/post/qxlihdtknfobguubugym@forum.dlang.org):
+```
+int f(int a)
+  in assert(a>0)
+  out(r) assert(b==a+1)
+{
+  return a+1;
+}
+```
+#### 5. [imply `assert` in addition to allowing contracts as naked expressions](http://forum.dlang.org/post/mailman.2337.1494961562.31550.digitalmars-d@puremagic.com):
+```
+int foo(T, U)(T t, U u)
+if (sigConstraints!T && sigConstraints!U)
+in (t > 0 && u < 10)
+out(foo > 1 && foo < 5 )
```

Comment:
That's actually why I'm not a fan of this syntax. It seemed like straightforward statements were ready-made and I could avoid such issues by using them instead. It doesn't seem too high a price, at the cost of a mere semicolon.